Docs build workflow

UmairK5669 opened the pull request:
**Description:**
Added a workflow to build the sphinx docs and check for any errors in the build. These errors will be annotated automatically to the PR. 


**Related Issues:**
#16 

**Tips for the Reviewer:**
Commit changes to docs build and observe workflow, if the changes break the build process there will be annotations appearing. 

Also, we should probably update the readme.md or coding.md to guide contributors on what to do when a check fails (check annotations etc)

**Checklist:**
Please be sure to check all boxes honestly. This is to ensure a smooth development process, and to reduce the 
likelihood of needing to make additional changes later on.
- [X] I understand that changes authored by individuals who have not signed a Contributor License Agreement (CLA), 
or whose authorship we cannot verify, may not be accepted.
- [X] These contribution address a triaged issue.
- [X] I have performed a self-review of my code and my changes generate no new warnings.
- [X] I have included unit tests for my code contributions, and all tests are passing.
- [X] The docstrings are complete, include doctests demonstrating usage, and are properly formatted. 
I have verified that any updates to the project documentation are complete and look okay.
- [X] I have confirmed all my code contributions are formatted in accordance with the project's Flake8, Black, 
and isort style configurations.
- [X] No unnecessary changes have been made to the Poetry lock file, but `pyproject.toml` and the Poetry lock file 
have been updated to reflect any dependency changes.
- [X] I acknowledge that upon the submission of this pull request, Qoherent will assume the copyright for this 
contribution.

Note: Sorry for the polluted PR, unsure why that happened. There should only be the Docs build workflow here, the codecov one and flake8 should be in a separate PR and should not be merged in yet. 

<!-- A heartfelt thank you from everyone at Qoherent and the broader radio community for taking the time to 
contribute to RIA Core. 🙏💖 -->

mrl280 commented on pyproject.toml:
```diff
@@ -75,6 +76,7 @@ exclude = '''
     /(
           \.git
         | \.github
+        | .tox
```
Looking at the patterns above, should this be `\.tox`?

mrl280 commented on .github/workflows/format.yml:
```diff
@@ -0,0 +1,43 @@
+            poetry lock --no-update
+            poetry install
+            pip install black
+            pip install isort
```
Shouldn't Black and isort come with `poetry install --with dev`?

mrl280 commented on .github/workflows/build-codecov.yml:
```diff
@@ -0,0 +1,43 @@
+          poetry lock --no-update
+          poetry install --with test
+          pip install pytest
+          pip install pytest-cov
```
Shouldn't pytest and pytest-cov come with `poetry install --with test`?

mrl280 commented on .github/codecov.yml:
Should this file be in the project root?

![image](https://github.com/qoherent/ria/assets/81438578/c4902df4-1501-49d8-bc0c-7d52be98fb6e)
Source: https://docs.codecov.com/docs/codecov-yaml#repository-yaml

mrl280 reviewed (commented):
Great work! These workflows are going to be very valuable. I've added a couple of comments/questions inline.

mrl280 commented:
Note: Let's remember to squash these commits at merge time.

UmairK5669 commented:
> Should this be tox.ini?

I was running flake8 locally and it was lining through .tox as well which was causing issues however, I realize now it wouldn't get pushed to GitHub so no need to keep it in. 

> Shouldn't Black and isort come with poetry install --with dev?

> Shouldn't pytest and pytest-cov come with poetry install --with test?

Unfortunately, using poetry doesn't seem to work for our workflows. I know it is possible since I have seen others do it but I can't figure out why it doesn't work. Check out [this commit](https://github.com/qoherent/ria/pull/26/commits/e7ff37bb8115e931d57f429c748e6ec80a887ec3) I made on the flake8-workflow PR, the [workflow](https://github.com/qoherent/ria/actions/runs/8412504943/job/23033431731) that ran as a result of this changed failed because the command line black wasn't found. Have tried this without the `--with dev` argument as well and it still doesn't work. I think we can merge in for now, report it as a bug and come back to this afterwards. 

> Should this file be in the project root?

Yes it should! Fixed. 

Also, I added an artifact that gets produced when the docs build workflow runs so now, we can see all the updates directly through GitHub without having to fetch the code and build the docs ourselves. I'll demo this in tomorrow's meeting for further clarity.

mrl280 reviewed (approved):
Looks good, let's remember to squash the commits
